File: src/main/java/com/uiptv/db/SeriesEpisodeDb.java
package com.uiptv.db;

import com.uiptv.model.Account;
import com.uiptv.model.Channel;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import static com.uiptv.db.DatabaseUtils.DbTable.SERIES_EPISODE_TABLE;
import static com.uiptv.db.DatabaseUtils.insertTableSql;
import static com.uiptv.db.SQLConnection.connect;

public class SeriesEpisodeDb extends BaseDb {
    private static SeriesEpisodeDb instance;

    public SeriesEpisodeDb() {
        super(SERIES_EPISODE_TABLE);
    }

    public static synchronized SeriesEpisodeDb get() {
        if (instance == null) {
            instance = new SeriesEpisodeDb();
        }
        return instance;
    }

    public List<Channel> getEpisodes(Account account, String categoryId, String seriesId) {
        return getAll(" WHERE accountId=? AND categoryId=? AND seriesId=?", new String[]{account.getDbId(), safeCategoryId(categoryId), seriesId});
    }

    public boolean isFresh(Account account, String categoryId, String seriesId, long maxAgeMs) {
        String sql = "SELECT MAX(cachedAt) FROM " + SERIES_EPISODE_TABLE.getTableName() + " WHERE accountId=? AND categoryId=? AND seriesId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, account.getDbId());
            statement.setString(2, safeCategoryId(categoryId));
            statement.setString(3, seriesId);
            try (ResultSet rs = statement.executeQuery()) {
                if (rs.next()) {
                    long cachedAt = rs.getLong(1);
                    return cachedAt > 0 && (System.currentTimeMillis() - cachedAt) <= maxAgeMs;
                }
            }
        } catch (SQLException ignored) {
        }
        return false;
    }

    public void saveAll(Account account, String categoryId, String seriesId, List<Channel> episodes) {
        deleteBySeries(account.getDbId(), categoryId, seriesId);
        long cachedAt = System.currentTimeMillis();
        for (Channel channel : episodes) {
            insert(account, categoryId, seriesId, channel, cachedAt);
        }
    }

    public void deleteByAccount(String accountId) {
        String sql = "DELETE FROM " + SERIES_EPISODE_TABLE.getTableName() + " WHERE accountId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute delete query", e);
        }
    }

    private void deleteBySeries(String accountId, String categoryId, String seriesId) {
        String sql = "DELETE FROM " + SERIES_EPISODE_TABLE.getTableName() + " WHERE accountId=? AND categoryId=? AND seriesId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.setString(2, safeCategoryId(categoryId));
            statement.setString(3, seriesId);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute delete query", e);
        }
    }

    private void insert(Account account, String categoryId, String seriesId, Channel channel, long cachedAt) {
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(insertTableSql(SERIES_EPISODE_TABLE))) {
            statement.setString(1, account.getDbId());
            statement.setString(2, safeCategoryId(categoryId));
            statement.setString(3, seriesId);
            statement.setString(4, channel.getChannelId());
            statement.setString(5, channel.getName());
            statement.setString(6, channel.getCmd());
            statement.setString(7, channel.getLogo());
            statement.setString(8, channel.getSeason());
            statement.setString(9, channel.getEpisodeNum());
            statement.setString(10, channel.getDescription());
            statement.setString(11, channel.getReleaseDate());
            statement.setString(12, channel.getRating());
            statement.setString(13, channel.getDuration());
            statement.setString(14, channel.getExtraJson());
            statement.setLong(15, cachedAt);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute insert query", e);
        }
    }

    @Override
    Channel populate(ResultSet resultSet) {
        Channel channel = new Channel();
        channel.setDbId(nullSafeString(resultSet, "id"));
        channel.setChannelId(nullSafeString(resultSet, "channelId"));
        channel.setName(nullSafeString(resultSet, "name"));
        channel.setCmd(nullSafeString(resultSet, "cmd"));
        channel.setLogo(nullSafeString(resultSet, "logo"));
        channel.setSeason(nullSafeString(resultSet, "season"));
        channel.setEpisodeNum(nullSafeString(resultSet, "episodeNum"));
        channel.setDescription(nullSafeString(resultSet, "description"));
        channel.setReleaseDate(nullSafeString(resultSet, "releaseDate"));
        channel.setRating(nullSafeString(resultSet, "rating"));
        channel.setDuration(nullSafeString(resultSet, "duration"));
        channel.setExtraJson(nullSafeString(resultSet, "extraJson"));
        return channel;
    }

    private String safeCategoryId(String categoryId) {
        return categoryId == null ? "" : categoryId.trim();
    }
}

File: server/src/main/kotlin/com/uiptv/db/SeriesEpisodeDb.kt
package com.uiptv.db

import com.uiptv.model.Account
import com.uiptv.model.Channel
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SortOrder
import org.jetbrains.exposed.sql.Table
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.transaction

class SeriesEpisodeDb private constructor() {
    companion object {
        private val instance = SeriesEpisodeDb()
        private const val EMPTY = ""

        @JvmStatic
        fun get(): SeriesEpisodeDb = instance
    }

    fun getEpisodes(account: Account, categoryId: String?, seriesId: String): List<Channel> =
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.selectAll()
                .where {
                    (SeriesEpisodeTable.accountId eq (account.dbId ?: EMPTY)) and
                        (SeriesEpisodeTable.categoryId eq safeCategoryId(categoryId)) and
                        (SeriesEpisodeTable.seriesId eq seriesId)
                }
                .map(ResultRow::toSeriesEpisodeChannel)
        }

    fun getEpisodesFromFreshestCategory(account: Account, seriesId: String): List<Channel> {
        val freshestCategoryId = getFreshestCategoryId(account, seriesId) ?: return emptyList()
        return getEpisodes(account, freshestCategoryId, seriesId)
    }

    fun isFresh(account: Account, categoryId: String?, seriesId: String, maxAgeMs: Long): Boolean {
        if (maxAgeMs <= 0) {
            return false
        }
        return try {
            latestCachedAtForCategory(account, categoryId, seriesId)?.let { cachedAt -> isFreshEnough(cachedAt, maxAgeMs) } ?: false
        } catch (_: Exception) {
            false
        }
    }

    fun isFreshInAnyCategory(account: Account, seriesId: String, maxAgeMs: Long): Boolean {
        if (maxAgeMs <= 0) {
            return false
        }
        return try {
            latestCachedAtInAnyCategory(account, seriesId)?.let { cachedAt -> isFreshEnough(cachedAt, maxAgeMs) } ?: false
        } catch (_: Exception) {
            false
        }
    }

    fun saveAll(account: Account, categoryId: String?, seriesId: String, episodes: List<Channel>) {
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.deleteWhere {
                (accountId eq (account.dbId ?: EMPTY)) and
                    (this.categoryId eq safeCategoryId(categoryId)) and
                    (this.seriesId eq seriesId)
            }
            val cachedAt = System.currentTimeMillis()
            episodes.forEach { channel ->
                SeriesEpisodeTable.insert { row ->
                    row[accountId] = account.dbId
                    row[this.categoryId] = safeCategoryId(categoryId)
                    row[this.seriesId] = seriesId
                    row[channelId] = channel.channelId
                    row[name] = channel.name
                    row[cmd] = channel.cmd
                    row[logo] = channel.logo
                    row[season] = channel.season
                    row[episodeNum] = channel.episodeNum
                    row[description] = channel.description
                    row[releaseDate] = channel.releaseDate
                    row[rating] = channel.rating
                    row[duration] = channel.duration
                    row[extraJson] = channel.extraJson
                    row[this.cachedAt] = cachedAt
                }
            }
        }
    }

    fun deleteByAccount(accountId: String) {
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.deleteWhere { this.accountId eq accountId }
        }
    }

    private fun latestCachedAtForCategory(account: Account, categoryId: String?, seriesId: String): Long? =
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.selectAll()
                .where {
                    (SeriesEpisodeTable.accountId eq (account.dbId ?: EMPTY)) and
                        (SeriesEpisodeTable.categoryId eq safeCategoryId(categoryId)) and
                        (SeriesEpisodeTable.seriesId eq seriesId)
                }
                .orderBy(SeriesEpisodeTable.cachedAt to SortOrder.DESC)
                .limit(1)
                .firstOrNull()
                ?.get(SeriesEpisodeTable.cachedAt)
        }

    private fun latestCachedAtInAnyCategory(account: Account, seriesId: String): Long? =
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.selectAll()
                .where {
                    (SeriesEpisodeTable.accountId eq (account.dbId ?: EMPTY)) and
                        (SeriesEpisodeTable.seriesId eq seriesId)
                }
                .orderBy(SeriesEpisodeTable.cachedAt to SortOrder.DESC)
                .limit(1)
                .firstOrNull()
                ?.get(SeriesEpisodeTable.cachedAt)
        }

    private fun getFreshestCategoryId(account: Account, seriesId: String): String? =
        transaction(SqlConnectionRuntime.database()) {
            SeriesEpisodeTable.selectAll()
                .where {
                    (SeriesEpisodeTable.accountId eq (account.dbId ?: EMPTY)) and
                        (SeriesEpisodeTable.seriesId eq seriesId)
                }
                .orderBy(
                    SeriesEpisodeTable.cachedAt to SortOrder.DESC,
                    SeriesEpisodeTable.id to SortOrder.DESC
                )
                .limit(1)
                .firstOrNull()
                ?.get(SeriesEpisodeTable.categoryId)
                ?.trim()
                ?.ifEmpty { EMPTY }
        }

    private fun isFreshEnough(cachedAt: Long, maxAgeMs: Long): Boolean =
        cachedAt > 0 && (System.currentTimeMillis() - cachedAt) <= maxAgeMs

    private fun safeCategoryId(categoryId: String?): String = categoryId?.trim().orEmpty()
}

private object SeriesEpisodeTable : Table(DatabaseUtils.DbTable.SERIES_EPISODE_TABLE.tableName) {
    val id = integer("id").autoIncrement()
    val accountId = text("accountId").nullable()
    val categoryId = text("categoryId").nullable()
    val seriesId = text("seriesId").nullable()
    val channelId = text("channelId").nullable()
    val name = text("name").nullable()
    val cmd = text("cmd").nullable()
    val logo = text("logo").nullable()
    val season = text("season").nullable()
    val episodeNum = text("episodeNum").nullable()
    val description = text("description").nullable()
    val releaseDate = text("releaseDate").nullable()
    val rating = text("rating").nullable()
    val duration = text("duration").nullable()
    val extraJson = text("extraJson").nullable()
    val cachedAt = long("cachedAt").nullable()

    override val primaryKey = PrimaryKey(id)
}

private fun ResultRow.toSeriesEpisodeChannel(): Channel =
    Channel(
        dbId = this[SeriesEpisodeTable.id].toString(),
        channelId = this[SeriesEpisodeTable.channelId],
        name = this[SeriesEpisodeTable.name],
        cmd = this[SeriesEpisodeTable.cmd],
        logo = this[SeriesEpisodeTable.logo],
        season = this[SeriesEpisodeTable.season],
        episodeNum = this[SeriesEpisodeTable.episodeNum],
        description = this[SeriesEpisodeTable.description],
        releaseDate = this[SeriesEpisodeTable.releaseDate],
        rating = this[SeriesEpisodeTable.rating],
        duration = this[SeriesEpisodeTable.duration],
        extraJson = this[SeriesEpisodeTable.extraJson]
    )
